nomenklatura/store/sql.py

```diff
@@ -0,0 +1,167 @@
+                proxy = self.assemble(current_stmts)
+                if proxy is not None:
+                    yield proxy
+                current_id = entity_id
+                current_stmts = []
+            current_stmts.append(stmt)
+        if len(current_stmts):
+            proxy = self.assemble(current_stmts)
+            if proxy is not None:
+                yield proxy
+
+
+class SqlWriter(Writer[DS, CE]):
+    BATCH_STATEMENTS = 10_000
+
+    def __init__(self, store: SqlStore[DS, CE]):
+        self.store: SqlStore[DS, CE] = store
+        self.batch: Optional[Set[Statement]] = None
+        self.batch_size = 0
+        self.insert = get_upsert_func(self.store.engine)
+
+    def flush(self) -> None:
+        if self.batch:
+            values = [pack_sql_statement(s) for s in self.batch]
+            istmt = self.insert(self.store.table).values(values)
+            stmt = istmt.on_conflict_do_update(
+                index_elements=["id"],
+                set_=dict(
+                    canonical_id=istmt.excluded.canonical_id,
+                    schema=istmt.excluded.schema,
+                    prop_type=istmt.excluded.prop_type,
+                    target=istmt.excluded.target,
+                    lang=istmt.excluded.lang,
+                    original_value=istmt.excluded.original_value,
+                    last_seen=istmt.excluded.last_seen,
+                ),
+            )
+            with self.store.engine.connect() as conn:
+                conn.begin()
+                conn.execute(stmt)
+                conn.commit()
+        self.batch = set()
+        self.batch_size = 0
+
+    def add_statement(self, stmt: Statement) -> None:
+        if self.batch is None:
+            self.batch = set()
+        if stmt.entity_id is None:
+            return
+        if self.batch_size >= self.BATCH_STATEMENTS:
+            self.flush()
+        canonical_id = self.store.resolver.get_canonical(stmt.entity_id)
+        stmt.canonical_id = canonical_id
+        self.batch.add(stmt)
+        self.batch_size += 1
+
+    def pop(self, entity_id: str) -> List[Statement]:
+        self.flush()
+        table = self.store.table
+        q = select(table).where(table.c.entity_id == entity_id)
+        q_delete = delete(table).where(table.c.entity_id == entity_id)
+        statements: List[Statement] = list(self.store._iterate_stmts(q))
+        with self.store.engine.connect() as conn:
+            conn.begin()
+            conn.execute(q_delete)
+            conn.commit()
+        return statements
+
+
+class SqlView(View[DS, CE]):
+    def __init__(
+        self, store: SqlStore[DS, CE], scope: DS, external: bool = False
+    ) -> None:
+        super().__init__(store, scope, external=external)
+        self.store: SqlStore[DS, CE] = store
+
+    def get_entity(self, id: str) -> Optional[CE]:
+        table = self.store.table
+        ids = [str(i) for i in self.store.resolver.connected(id)]
```

Comment:
Ah, interesting, so the idea here is that we're not believing the `stmt.canonical_id` in the table? While that works, it's a a bit different from all the other store implementations we have now....

Reply:
hm, no, i guess i just misunderstood your comment here: https://github.com/opensanctions/nomenklatura/pull/125#discussion_r1289624836